### ISSUE TITLE
expand single darkroom module preference when changing group

### ISSUE THREAD

**Issue by elstoc** [darktable-org/darktable#5129]
If you have the "expand single darkroom module at a time" preference checked this preference is not correctly observed when you change module groups.

If, for example, I perform the following steps

1. Expand and enable the 'exposure' module in the Basic group
2. Expand and enable the 'tone curve' module in the Tone group
3. Select the Active Modules group

Now, in the Active Modules group, both the 'exposure' and the 'tone curve' modules are expanded. That's not "expanding a single darkroom module at a time"! If I want to go back to only one module being expanded I have to collapse one of the expanded modules and then re-expand it. 

Using variations on this approach, it's fairly easy to get multiple modules to display in any of the groups. Similar issues are found if you click twice on a group to view all modules or use the search functionality.

I hesitate to just "fix" the issue, since it does look like it was intended by _someone_: ``imageop.c`` has

```
      if(m != module && dt_iop_shown_in_group(m, current_group))
      {
        all_other_closed = all_other_closed && !m->expanded;
        dt_iop_gui_set_single_expanded(m, FALSE);
      }
```

So the question is, what is the expected (or desired) behaviour here? Either it should be 'only one module expanded, regardless of group' or it should be 'one module may be expanded in each group'. But currently it's not quite one or the other.

My preference is for the first option - each time you expand a module, all other modules should collapse, in every group. It's a simple fix and it's behaviour that's easily understood and consistent with the preference description.

For the second option (one module per group), we would have to treat each group independently (probably regardless of the expansion preference setting) and have some way to record, for each group, which module instance(s) are expanded (in a multi-instance scenario just recording the module so name won't do), possibly also independently recording the scroll position in each (See #3085) and gracefully handling changes to the history stack. Then each time the user moves to a new group, reload the last state.

**Comment by johnny-bit** [comment]
That particular code is member of "using shift click to expand more than one module despite single module expand being set"... and that leads to host of problems. you can view git-blame for that section of code, maybe history from it would shade some light on the issue.

so... that should be very complex handling on all corner cases :/ and seems like 2nd option would be the handling needed including respecting shift-click on active/favourite modules list...

**Comment by elstoc** [comment]
Well essentially shift-click just temporarily reverses the behaviour set in the preference. So if you've checked the preference (expand single module) then shift-click will expand a module without collapsing the others. If you've unchecked the preference then shift-click will expand the clicked module and collapse the others (in that group).

So the second option would allow for "expand single module" (either via click with the option checked or shift-click otherwise) to instead mean "expand single module in each group" but would also mean that any expansion would be local to the group in which it is expanded, regardless of the preferences or shift-click behaviour. So if I expanded exposure in the favourites, it wouldn't also be expanded in the tone group.

I'm more than happy to just change the functionality that collapses everything else so that it does so globally (option 1). Then the alternative (don't collapse anything else when expanding) would continue to work as now. Perhaps I should do that as a short-term fix and leave this issue open to discuss longer-term enhancements.

**Comment by johnny-bit** [comment]
(personally I'm very in favour of option 1)

**Comment by junkyardsparkle** [comment]
Should we re-visit this now that the module groups are easily user-configurable? Previously, the "option one" described here probably would have kept me from using single-expanded-module mode (which is ok). Now, with the option of having as many groups as desired, and being able to switch between setups with presets, I guess it's time for me to stop thinking of the groups as anything like "tabs" (which was kind of the basis for #3085). My thought is that if they are decisively non-tabs, maybe their appearance could be subtly shifted towards "radio button" to make that more obvious.

I was thinking of closing #3085, but I'll leave it for now since there still seems to be some discussion on the table... just want to say that I'm personally re-evaluating all of that in light of recent development. :-)

**Comment by elstoc** [comment]
I've retested this on current master and it is still an issue. You could say it's even more of an issue now, since a module can now appear in any number of module groups (it could appear in all of them). I think the only way forward is to proceed with my PR #5155 but with a preference option to allow the current behaviour to be retained.

**Comment by junkyardsparkle** [comment]
>  You could say it's even more of an issue now, since a module can now appear in any number of module groups (it could appear in all of them).

Fair point. I'm still figuring out how to make the best use of the added configurability. The way it works now pushes me towards larger numbers of groups, each with unique module lists.

**Comment by elstoc** [comment]
The more groups you have the more stuff will be expanded if you ever need to use the 'active modules' group.

**Comment by elstoc** [comment]
Ok I think this is basically resolved by #5155 so closing
